In [1]:
import pandas as pd
from sklearn.cluster import KMeans

In [2]:
df_scaled = pd.read_csv('../data/valorant_scaled_data.csv')
X_train = df_scaled.drop(columns=['player_id', 'archetype'])

In [3]:
# Train the K-Means model on the unlabeled features
kmeans = KMeans(n_clusters=5, random_state=42, n_init=10)
discovered_clusters = kmeans.fit_predict(X_train)

In [4]:
# Load the original unscaled data so we can interpret the raw numbers of the new clusters
df_original = pd.read_csv('../data/valorant_synthetic_data_raw.csv')

In [7]:
# Assign the model's discovered cluster numbers back to the original unscaled dataframe
df_original['assigned_cluster'] = discovered_clusters

In [8]:
# Calculate the average stats for each discovered cluster to understand the playstyles it found
cluster_profiles = df_original.groupby('assigned_cluster').mean(numeric_only=True)
display(cluster_profiles)

,adr,kast_percent,first_bloods_per_round,first_deaths_per_round,assists_per_round,clutch_win_rate,survival_rate,trade_percent,headshot_percent,econ_rating,sniper_kill_percent
assigned_cluster,,,,,,,,,,,
0,148.973943,71.996556,0.079257,0.051764,0.205693,24.948684,29.645244,8.023282,28.208560,75.004687,7.892237
1,135.717045,71.895176,0.109067,0.107157,0.298426,15.241449,24.951402,14.014408,21.124842,62.034799,10.297090
2,165.365577,70.306077,0.180479,0.179593,0.148932,11.759593,17.811853,10.030005,22.091285,54.817986,24.819196
3,124.868634,78.450326,0.063478,0.062587,0.453472,15.236744,35.194696,21.844885,20.218913,64.775077,5.546212
4,110.311215,74.488958,0.049854,0.081637,0.540272,9.961062,28.218014,15.070853,15.872923,59.615147,3.304468


### Interpreting the Discovered Clusters
In a real-world scenario, you look at the output of Cell 6 to name the clusters yourself. 

For example, if the output shows that `Cluster 0` has an average `first_bloods_per_round` of 0.18 and a `sniper_kill_percent` of 25%, you would map `Cluster 0` to the string "Aggressive Entry" in your backend. If `Cluster 2` has the highest `assists_per_round` and lowest `adr`, you map it to "Utility-Heavy".

This proves the model can successfully take blind data and organize players into distinct, recognizable Valorant playstyles.